### PR TITLE
Declarative configuration schema validation

### PR DESCRIPTION
**Pull request by kennsippell**
This is an associated change from medic/cht-core#5657.

One key impact from this feature is the ability to write impact queries to understand how tasks are behaving on users phones. In previously versions of the declarative configuraiton, we default to using the "index" of a task, target, or event. This is problematic, because it will lead to data integrity issues if elements are ever reorderd or removed from within the configuration.

This change adds two additional constraints to the declarative configuration schema:
* `Task.name` is now a required unique attribute
* `Task.event[n].id` is now a required unique attribute when there is more than one event
* `Target.id` although always required, is now a required unique attribute

There is currently custom validation for targets and no validation for tasks. This change removes the custom validation for targets and replaces it with full schema validation based on [Joi](https://hapi.dev/family/joi/?v=16.1.8).

Example updates:
* Muso - https://github.com/medic/medic-projects/pull/7082
* Cht-Reference-App - https://github.com/medic/cht-core/pull/6155

### CODE REVIEW

**Comment by kennsippell** [comment]
One thing we might consider adding here is the use of [descriptions](https://hapi.dev/family/joi/?v=16.1.8#anydescriptiondesc). I love the interface to use Joi, but the errors it is outputting are very direct and users may benefit if we duplicate the documentation to be directly in code also. I didn't do that in the v1 here.

File: src/lib/validate-declarative-schema.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
+const path = require('path');
+const Joi = require('@hapi/joi');
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/validate-declarative-schema.js]: but stackoverflow said to ...

File: src/lib/validate-declarative-schema.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  .unique('name')
+  .required();
+
+const validate = (logEvent, projectDir, filename, schema) => {
+  const pathToTasks = path.join(projectDir, filename);
+  let tasks;
+  try {
+    tasks = require(pathToTasks);
+  } catch (err) {
+    return `Failed to parse file ${pathToTasks}. ${err}`;
```
[the remark this answers is not in the document]
Comment (kennsippell) [reply src/lib/validate-declarative-schema.js]: good catch thanks

**Comment by kennsippell** [comment]
> I wonder if we should include a target-version argument so you could work offline so long as you know what version of the core you'll be deploying against eventually...

In general, I think it makes sense to encourage authors of partner code to follow one schema for declarative configuration instead of introducing a versioned schema -- it's very hard to document and understand for new users. Although it is true that these name/id values aren't technically required for the config to function in previous versions of the webapp, 1) it's fairly straightforward to add them, 2) authors can workaround temporarily with --debug, 3) it reduces the odds of a production issue during upgrade, 4) if necesary, they can continue to use an old version of medic-conf until they're ready.

**Comment by kennsippell** [comment]
#276

**Comment by dianabarsan** [comment]
Hey @kennsippell 
I've been attempting some AT on this and have some comments:

### Tasks:

When I have 2 tasks with the same name, I get this error:
```
ERROR tasks.js invalid schema: "[1]" contains a duplicate value
```
It's really unclear what the "duplicate value" is. Initially I believed it was a duplicate event id. 
Looking at Joi documentation, the error it sends has information about the duplicate in the `details` property: https://hapi.dev/family/joi/?v=16.1.8#arrayunique
We could easily improve the message to look something like:
```
"[1]" contains a duplicate value for the "name" field: "some value"
```
The details property is an array, with one item per validation error. 


When I'm using an invalid `priority`, I get this error:
```
"[1].priority" must be one of [object]
```
It looks like `joi.function` in alternatives is displayed as "object" in errors, cause I see the same behavior for `contactLabel`. 
If my `contactLabel` is `contactLabel: { a: 1 },` the message I get is:
```
ERROR tasks.js invalid schema: "[0].contactLabel" must be one of [string, object] 
```
It's confusing since I *was* using an object, but in fact it should be a function.

Should we allow empty `appliesToType` arrays? 


If I create an `event` that has both `dueDate` and `days`, I get this error:
```
"[0].events[0].days" is not allowed. "[0].events[0].dueDate" is not allowed
```
Assuming I haven't read the validation code and don't know that they are actually mutually exclusive, and remove both, I get this error:
```
"[0].events[0].days" is required. "[0].events[0].dueDate" is required 
```
Pretty confusing, ha? 


### Targets:

Should we allow empty `appliesToType` arrays? 


Using an incorrect `date` (for example: `{ a: 1 }`), I get this error:
```
"[0].date" does not match any of the allowed types.
```
Using another incorrect date (for example: `not a date`):
```
 "[0].date" must be one of [reported, now, object].
 ```
 
 When I have only one target, and it's invalid, I get an additional error:
 ```
 "value" does not contain 1 required value(s)
 ```
 Maybe we should remove the `required()` from the object and add a `min(1)` to the array ? This way the error will be
 ```
"value" must contain at least 1 items 
```
and it will only appear when there are no targets. 
Is this desired though? Is it mandatory to have at least one target?
 

Would you mind also resolve conflicts, please?

**Comment by kennsippell** [comment]
@dianabarsan Thanks for the feedback. I've done a bit of work to improve the error reporting both generically and for the specific scenarios you called out above. Would you take another look?

> Maybe we should remove the required() from the object and add a min(1) to the array ?

I don't think this is a good idea, since the user shouldn't need to have tasks to compile their targets and vice versa.

Take a look at the new error messages and let me know what you think.

**Comment by dianabarsan** [comment]
Hey @kennsippell , looking good! I took another look and have a couple of comments left. 

1. I am still seeing an error when I have no targets. 
```
ERROR targets.js invalid schema: 
ERROR "value" does not contain 1 required value(s). Value is: "" 
```
when my `targets.js` is:
```
module.exports = [];
```

2. The target validation message for `passesIf` mentions `groupBy`. However, there is no definition for `groupBy`. Were you meaning to add it?
Also, we could tweak the error message to detail what type `passesIf` accepts, currently it just mentions whether or not it's required.

**Comment by kennsippell** [comment]
Oops! The group by comment only makes sense with this change! They are both going into 3.1. Do u mind if it keep it as is?
#277

**Comment by kennsippell** [comment]
@dianabarsan Pushed a fix for 2. This should be good to merge and we can address the documentation feedback in a new PR.  

I've avoided documenting `emitCustom` because it is hard to explain. Not saying that isn't valuable there, but it is a fairly complex and advanced outlier.

**Pull request merged**